python/ray/data/_internal/planner/exchange/push_based_shuffle_task_scheduler.py

```diff
@@ -621,39 +706,65 @@ def _compute_shuffle_schedule(
-            node_parallelism = min(
-                num_cpus, num_input_blocks // len(num_cpus_per_node_map)
+            # First find how many merge tasks we should run on this node.
+            # We take the min of the number of CPUs on this node and the number
+            # of input blocks that we haven't scheduled yet, in case there are
+            # fewer input blocks than CPU slots on this node.
+            num_cpu_slots = min(num_cpus, num_input_blocks_remaining)
+            num_merge_tasks_on_cur_node = int(
+                num_cpu_slots // num_tasks_per_map_merge_group
             )
-            num_merge_tasks = node_parallelism // num_tasks_per_map_merge_group
-            for i in range(num_merge_tasks):
+            # For small datasets, the number of tasks to run may be less than
+            # the total CPU slots available.
+            num_merge_tasks_on_cur_node = min(
+                num_merge_tasks_on_cur_node, num_total_merge_tasks
+            )
+            for i in range(num_merge_tasks_on_cur_node):
                 merge_task_placement.append(node)
-            num_merge_tasks_per_round += num_merge_tasks
+                num_input_blocks_remaining -= merge_factor
+            num_merge_tasks_per_round += num_merge_tasks_on_cur_node
 
             # Handle the case where a single node cannot fit a group of map and
             # merge tasks, but we can spread the group across multiple distinct
             # nodes.
-            leftover_cpus += node_parallelism % num_tasks_per_map_merge_group
-            if num_merge_tasks == 0 and leftover_cpus > num_tasks_per_map_merge_group:
+            leftover_cpus += num_cpus % num_tasks_per_map_merge_group
+            if leftover_cpus >= num_tasks_per_map_merge_group:
                 merge_task_placement.append(node)
                 num_merge_tasks_per_round += 1
                 leftover_cpus -= num_tasks_per_map_merge_group
+                num_input_blocks_remaining -= merge_factor
+
         if num_merge_tasks_per_round == 0:
-            merge_task_placement.append(list(num_cpus_per_node_map)[0])
-            num_merge_tasks_per_round = 1
+            # For small datasets, make sure we have at least one merge task.
+            for node, num_cpus in num_cpus_per_node_map.items():
+                if num_cpus >= 1:
+                    merge_task_placement.append(node)
+                    num_merge_tasks_per_round = 1
+                    break
 
         assert num_merge_tasks_per_round == len(merge_task_placement)
-        num_map_tasks_per_round = max(task_parallelism - num_merge_tasks_per_round, 1)
+        assert num_merge_tasks_per_round > 0, num_merge_tasks_per_round
+        # Use the remaining CPUs to execute map tasks.
+        num_map_tasks_per_round = task_parallelism - num_merge_tasks_per_round
+        # Make sure there is at least one map task in each round.
+        num_map_tasks_per_round = max(num_map_tasks_per_round, 1)
```

Comment:
should we also decrement num_merge_tasks_per_round by 1 in this case?

Reply:
It should be okay to have num_map_tasks_per_round + num_merge_tasks_per_round > task_parallelism; this is a pretty narrow edge case anyway.

Reply:
sounds good. Just wanted to make sure no dead lock in this case.